DOC: simplify API index (#19727)
matplotlib/matplotlib: doc/api/index.rst

API
===

.. toctree::
   :hidden:

   api_changes

For recent changes, see :doc:`api_changes`.

When using the library you will typically create
:doc:`Figure <figure_api>` and :doc:`Axes <axes_api>` objects and 
call their methods to add content and modify the apprearance.

  - :doc:`figure_api`: axes creation, figure-level content
  - :doc:`axes_api`: most plotting methods, Axes labels, access to 
    axis styling, etc.  

Example: We create a Figure ``fig`` and Axes ``ax``. Then we call
methods on them to plot data, add axis labels and a figure title.
 
.. plot::
   :include-source:
   :align: center

   import matplotlib.pyplot as plt
   import numpy as np

   x = np.arange(0, 4, 0.1)
   y = np.sin(x*np.pi)

   fig, ax = plt.subplots(figsize=(3,2), constrained_layout=True)
   ax.plot(x, y)
   ax.set_xlabel('t [s]')
   ax.set_ylabel('S [V]')

>    fig.suptitle('Sine wave')

Reviewer: ```suggestion
   ax.set_title('Sine wave')
   fig.patch.set_color('lightsteelblue')
```
One would not use `fig.suptitle` with a single Axes. This looks a bit odd because the title is centered on the figure not the Axes. There are actually not too many reasonable uses for figure methods in this simple example. Coloring the background may be one of the better ones.

Author: Are we trying to soft deprecate `fig.set_facecolor`?  I understand not wanting duplicate ways of doing the same thing, but asking folks to modify a property like this is a discoverability issue.

Modules
-------

Alphabetical list of submodules:

.. toctree::

>    :maxdepth: 1

Reviewer: ```suggestion
   :maxdepth: 1
   :multicol-toc:
```
Maybe.
![grafik](https://user-images.githubusercontent.com/2836374/111854685-c1e68180-8920-11eb-9568-f5ec9cc7b9c3.png)

Author: Is this unreleased?  I get `/Users/jklymak/matplotlib/doc/api/index.rst:45: WARNING: Error in "toctree" directive:
unknown option: "multicol-toc".` on sphinx 3.5.3

Reviewer: Sorry, this should have been `:class: multicol-toc`.

`multicol-toc` is just a CSS class we have added to our `mpl.css`.

Author: oops reached the internet - didn't think to search our source code...

Reviewer: Sorry this does not seem to work. `.. toctree:` does not support a `:class:` attribute.

I've otherwise only used it for `.. contents:` so far. The above screenshot was a hacked CSS in the browser, because I didn't want to wait for a docs build.

Seems we have to live with the long list for now.

Author: Ha!  I thought I was going crazy...  Certainly adding a class directive to toctree seems like a good idea, but seems like an upstream issue.

API
===

.. toctree::
   :hidden:

   api_changes

For recent changes, see :doc:`api_changes`.

> Much of the library is accessed by creating :doc:`Figure <figure_api>` 
> and :doc:`Axes <axes_api>` objects and calling methods on them. 

Reviewer: ```suggestion
When using the library you will typically create
:doc:`Figure <figure_api>` and :doc:`Axes <axes_api>` objects and 
call their methods to add content and modify the apprearance.
```

>   - :doc:`figure_api`: axes creation, figure-level artists
>   - :doc:`axes_api`: most plotting methods, axes labels, access to 
>     axis styling, etc.  

Reviewer: ```suggestion
  - :doc:`figure_api`: axes creation, figure-level content
  - :doc:`axes_api`: most plotting methods, Axes labels, access to 
    axis styling, etc.  
```

> (Note that there is also a parallel `matplotlib.pyplot` API interface 
> that is considered useful for interactive work; see 
> :ref:`usage patterns <usage_patterns>`, below).

Reviewer: ```suggestion
Alternatively, there is also a parallel `matplotlib.pyplot` API interface 
that is considered useful for interactive work; see 
:ref:`usage patterns <usage_patterns>`, below.
```

Reviewer: But I'm not sure we should mention the distinction here at all. Particularly also because the following example uses the object-based approach, but still needs to use pyplot for Figure and Axes creation. It's quite confusing that the non-pyplot approach still needs pyplot. So, I'd not go into the details here.

Author: Sure - I struggled a bit with how to work this in, or if I should.  It is such a common point of confusion that someone wrote a whole section and stuck it at the top of the API docs, so it seemed at least referring to it near the top would be useful, but I'm happy to leave it out.

> Simple example: the ``ax`` object has an `.axes.Axes.plot` method 
> that is used to create the line plot.   

Reviewer: ```suggestion
Example: We create a Figure ``fig`` and Axes ``ax``. Then we call
methods on them to plot data, add axis labels and a figure title.
```

 
.. plot::
   :include-source:
   :align: center

   import matplotlib.pyplot as plt
   import numpy as np

   x = np.arange(0, 4, 0.1)
   y = np.sin(x*np.pi)

>    fig, ax = plt.subplots(figsize=(3,2), constrained_layout=True)

Reviewer: ```suggestion
   fig, ax = plt.subplots()
```
Keep it simple without parameters. While `constrained_layout` is nice, IMHO this is not the place to teach it (and the benefit for this simple plot is small).

Author: Yes, but then the figure is too large, in my opinion.

Reviewer: You could try with

~~~
.. plot::
   :scale: 50 %
~~~

but I'm not sure if that will look good.

Author: Yeah, I wish there was a two-column view....  We could just include the code and the figure as an image as well. I do think the example is useful, though, to make it concrete what we are talking about.

Author: By default, the ylabel is cutoff, so *some* formatting is necessary anyway..  Might as well keep what is above.  I don't think we need to explain every incantation everywhere it appears.

   ax.plot(x, y)
   ax.set_xlabel('t [s]')
   ax.set_ylabel('S [V]')
   fig.suptitle('Sine wave')

Modules
-------

> Alphabetical list of all submodules:

Reviewer: ```suggestion
Alphabetical list of submodules:
```
Does not carry relevant information. Shorter is better.

   toolkits/axes_grid1.rst
   toolkits/axisartist.rst
   toolkits/axes_grid.rst

.. _usage_patterns:

Usage patterns
--------------

Below we describe several common approaches to plotting with Matplotlib.

The pyplot API
^^^^^^^^^^^^^^

> `matplotlib.pyplot` is a collection of command style functions that make

Reviewer: ```suggestion
`matplotlib.pyplot` is a collection of functions that make
```
Maybe it's just me, but I have no idea what "command style functions" are. I feel it's still (or even better) understandable with only "functions".

Author: Sure, I just cut and pasted this...
